File: server/src/test/java/com/fluffy/exam/application/ExamImageServiceTest.java
package com.fluffy.exam.application;

import com.fluffy.auth.domain.Member;
import com.fluffy.auth.domain.MemberRepository;
import com.fluffy.exam.application.request.ExamImagePresignedUrlRequest;
import com.fluffy.exam.application.response.ExamImagePresignedUrlResponse;
import com.fluffy.exam.domain.Exam;
import com.fluffy.exam.domain.ExamImageRepository;
import com.fluffy.exam.domain.ExamRepository;
import com.fluffy.global.exception.ForbiddenException;
import com.fluffy.global.storage.response.PresignedUrlResponse;
import com.fluffy.global.web.Accessor;
import com.fluffy.support.AbstractIntegrationTest;
import com.fluffy.support.data.MemberTestData;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.web.multipart.MultipartFile;

class ExamImageServiceTest extends AbstractIntegrationTest {

    @Autowired
    private ExamImageService examImageService;

    @Autowired
    private MemberRepository memberRepository;

    @Autowired
    private ExamRepository examRepository;

    @Autowired
    private ExamImageRepository examImageRepository;

    @Test
    @DisplayName("시험 이미지에 대한 Presigend URL을 생성할 수 있다.")
    void createPresignedUrl() {
        Member member = createMember();
        Exam exam = createExam(member);

        String presignedUrl = "https://cdn.fluffy.run/exams/1/QWER-1234.png?X-Amz-BLABLA=1234";
        String fileUrl = "https://cdn.fluffy.run/exams/1/QWER-1234.png";
        when(storageClient.getPresignedUrl(any()))
                .thenReturn(new PresignedUrlResponse(presignedUrl, fileUrl));

        ExamImagePresignedUrlRequest request = new ExamImagePresignedUrlRequest("BLA-BLA.png", 1024L);
        Accessor accessor = new Accessor(member.getId());

        ExamImagePresignedUrlResponse response = examImageService.createPresignedUrl(exam.getId(), request, accessor);

        assertAll(
                () -> assertThat(response.presignedUrl()).isEqualTo(presignedUrl),
                () -> assertThat(response.imageUrl()).isEqualTo(fileUrl),
                () -> assertThat(examImageRepository.findAll()).hasSize(1)
        );
    }

    @Test
    @DisplayName("시험에 대한 작성자만 해당 Presigned URL을 생성할 수 있다.")
    void createPresignedUrlWithDifferentAuthor() {
        Member member = createMember();
        Member anotherMember = createAnotherMember();
        Exam exam = createExam(anotherMember);

        ExamImagePresignedUrlRequest request = new ExamImagePresignedUrlRequest("BLA-BLA.png", 1024L);
        Accessor accessor = new Accessor(member.getId());

        Long examId = exam.getId();
        assertThatThrownBy(() -> examImageService.createPresignedUrl(examId, request, accessor))
                .isInstanceOf(ForbiddenException.class);
    }

    private Member createMember() {
        Member member = MemberTestData.defaultMember().build();
        return memberRepository.save(member);
    }

    private Member createAnotherMember() {
        Member member = MemberTestData.defaultMember().withId(2L).build();
        return memberRepository.save(member);
    }

    private Exam createExam(Member member) {
        Exam exam = Exam.create("시험 제목", member.getId());
        return examRepository.save(exam);
    }

    private MultipartFile createMockMultipartFile() {
        return new MockMultipartFile(
                "image",
                "image.png",
                "image/png",
                "image".getBytes()
        );
    }
}

File: server/src/test/kotlin/com/fluffy/exam/application/ExamImageServiceTest.kt
package com.fluffy.exam.application

import com.fluffy.exam.application.request.ExamImagePresignedUrlRequest
import com.fluffy.exam.application.response.ExamImagePresignedUrlResponse
import com.fluffy.exam.domain.*
import com.fluffy.global.exception.ForbiddenException
import com.fluffy.global.exception.NotFoundException
import com.fluffy.global.storage.StorageClient
import com.fluffy.global.storage.response.PresignedUrlResponse
import com.fluffy.global.web.Accessor
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.unmockkObject
import org.springframework.data.repository.findByIdOrNull

class ExamImageServiceTest : BehaviorSpec({

    val storageClient = mockk<StorageClient>()
    val examRepository = mockk<ExamRepository>()
    val examImageRepository = mockk<ExamImageRepository>()
    val examImageService = ExamImageService(
        storageClient = storageClient,
        examRepository = examRepository,
        examImageRepository = examImageRepository
    )

    Given("createExamImage") {
        val examId = 1L
        val request = ExamImagePresignedUrlRequest(imageName = "image.jpg", fileSize = 1024L)
        val memberId = 1L
        val accessor = Accessor(memberId)

        When("시험이 없을 경우") {
            every { examRepository.findByIdOrNull(any()) } returns null

            Then("예외를 발생시킨다.") {
                shouldThrow<NotFoundException> {
                    examImageService.createExamImage(examId, request, accessor)
                }.message shouldBe "존재하지 않는 시험입니다. 시험 식별자: $examId"
            }
        }

        When("시험이 있을 경우") {
            val exam = mockk<Exam>()
            every { examRepository.findByIdOrThrow(any()) } returns exam

            And("시험 작성자가 아닐 경우") {
                every { exam.isNotWrittenBy(memberId) } returns true

                Then("예외를 발생시킨다.") {
                    shouldThrow<ForbiddenException> {
                        examImageService.createExamImage(examId, request, accessor)
                    }.message shouldBe "시험 작성자만 이미지를 업로드할 수 있습니다."
                }
            }

            And("시험 작성자일 경우") {
                mockkObject(ExamImage)

                val examImage = mockk<ExamImage>()
                val expected = "exam/$examId/UUID-1234-5678/image.jpg"
                every { examImage.path } returns expected
                every { ExamImage.create(any(), any(), any(), any()) } returns examImage

                every { exam.isNotWrittenBy(memberId) } returns false
                every { examImageRepository.save(any()) } returns examImage

                val actual = examImageService.createExamImage(examId, request, accessor)

                Then("시험 이미지를 생성하고, 경로를 반환한다.") {
                    actual shouldBe expected
                }

                unmockkObject(ExamImage)
            }
        }
    }

    Given("generatePresignedUrl") {
        val imagePath = "exam/1/UUID-1234-5678/image.jpg"

        When("이미지 경로를 받으면") {
            val presignedUrl = "https://s3-bucket.com/$imagePath"
            val fileUrl = "https://cdn.example.com/$imagePath"
            val response = PresignedUrlResponse(
                presignedUrl = presignedUrl,
                fileUrl = fileUrl
            )
            every { storageClient.getPresignedUrl(any()) } returns response

            val actual = examImageService.generatePresignedUrl(imagePath)
            val expected = ExamImagePresignedUrlResponse(
                presignedUrl = presignedUrl,
                imageUrl = fileUrl
            )

            Then("presignedUrl과 imageUrl을 반환한다.") {
                actual shouldBe expected
            }
        }
    }
})